### PR TITLE
BOLT 4, 9, 11: Base Atomic Multipath Payments.

### PR DESCRIPTION
**Pull request by ZmnSCPxj**
xref: https://lists.linuxfoundation.org/pipermail/lightning-dev/2018-November/001590.html

### CODE REVIEW

File: 04-onion-routing.md [hunk 1]
```diff
@@ -873,8 +931,54 @@ An _intermediate hop_ MUST NOT, but the _final node_:
   the final node's HTLC:
     - MUST return `final_incorrect_cltv_expiry` error.
   - if the `amt_to_forward` is greater than the `incoming_htlc_amt` from the
-  final node's HTLC:
+  final node's HTLC, AND `incomplete_payment` is not set:
     - MUST return a `final_incorrect_htlc_amount` error.
+  - if `incomplete_payment` is set:
+    - MUST use `amt_to_forward` to evaluate for
+      `incorrect_payment_amount` errors.
+    - SHOULD suspend incoming HTLCs:
```
Comment (renepickhardt) [inline 04-onion-routing.md]: Maybe change to : SHOULD suspend incoming HTLCs with the same payment hash:

File: 04-onion-routing.md [hunk 2]
```diff
@@ -257,6 +271,40 @@ compare its values against those of the HTLC. See the
...
+The sender of the payment MUST NOT use base atomic multipath if the
+`wait_on_incomplete` invoice feature is not set in the invoice.
+
+In a Base Atomic Multipath Payment, the sender MAY split the payment into
+two or more paths.
+The sender MAY also signal multipath while actually only using a single path.
+The same `payment_hash` is used in all paths.
+The final node hop data of each path
+MUST have the `incomplete_payment` flag set,
+and MUST have the same `amt_to_forward` for all paths.
```
[the remark this answers is not in the document]
Comment (renepickhardt) [reply 04-onion-routing.md]: Never mind saw that you put it to the requirements in line 9xx. Still it is not clear to me how to mitigate dos attacks (but I guess that did attack can also exist without base AMP)

File: 04-onion-routing.md [hunk 3]
```diff
@@ -873,8 +931,54 @@ An _intermediate hop_ MUST NOT, but the _final node_:
...
+arrived at the final node, and the final node claims the payment,
+then the final node has effectively sold the `payment_preimage` at a
+discount.
+
+This is effectively "economically rational atomicity".
+A final node that wants to lose money may voluntarily claim a partial
+payment that is less than its advertised payment amount.
+We assume that no final node would want to lose money.
+
+We assume here that the `payment_preimage` is of value.
```
Comment (renepickhardt) [inline 04-onion-routing.md]: I would leave out this 3rd Paragraph of the 4 added rational paragraphs. I think the other 3 work without this one

File: 04-onion-routing.md [hunk 4]
```diff
@@ -873,8 +931,54 @@ An _intermediate hop_ MUST NOT, but the _final node_:
...
+  - if `incomplete_payment` is set:
+    - MUST use `amt_to_forward` to evaluate for
+      `incorrect_payment_amount` errors.
+    - SHOULD suspend incoming HTLCs:
+      - if the incoming HTLC has an `amt_to_forward` that is
+        different from existing suspended payments of the same
+        `payment_hash`:
+        - MUST fail the HTLC.
+        - MUST return an `incorrect_payment_amount` error.
+      - if the total `incoming_htlc_amt` of all suspended HTLCs and
```
Comment (renepickhardt) [inline 04-onion-routing.md]: It is confusing. We are at the part where the amt_to_forward is different. Now you give an rational when to succeed the htlcs.

I would separate this somehow

File: 04-onion-routing.md [hunk 5]
```diff
@@ -791,6 +839,16 @@ The channel from the processing node has been disabled.
 
 The CLTV expiry in the HTLC is too far in the future.
 
+1. type: BADONION|PERM|22 (`invalid_flag`)
+
+An even-numbered bit of `flags0` is set, which is not understood
```
[the remark this answers is not in the document]
Comment (ZmnSCPxj) [reply 04-onion-routing.md]: In keeping, with "It's OK to be odd" rule, which has the corollary "It's BAD to be even".

File: 04-onion-routing.md (hunk 1 above)
Comment (ZmnSCPxj) [reply 04-onion-routing.md]: My mental model is a single-threaded server, so approximately, you would write code as below:

1.  If the `incomplete_payment` is set, suspend incoming HTLC (also set a timeout trigger to fail suspended HTLCs due to taking too long)
2.  Scan through all suspended HTLCs and sum up their total.
3.  If the total in all suspended HTLCs >= `amt_to_forward`, then claim them.

Although the changed wording seems fine to me, either way.

File: 04-onion-routing.md (hunk 4 above)
Comment (ZmnSCPxj) [reply 04-onion-routing.md]: They are on the same level, so my mental model is that these rules are evaluated in that order, and this rule is not gated by the `amt_to_forward` rule.

File: 04-onion-routing.md [hunk 6]
```diff
@@ -257,6 +271,40 @@ compare its values against those of the HTLC. See the
 If not for the above, since it need not forward payments, the final node could
 simply discard its payload.
 
+### Base Atomic Multipath Payments
```
[the remark this answers is not in the document]
Comment (ZmnSCPxj) [reply 04-onion-routing.md]: Atomic Multipath Payments is how the users know this feature, so I am hesitant to change it.

File: 04-onion-routing.md [hunk 7]
```diff
@@ -257,6 +271,40 @@ compare its values against those of the HTLC. See the
...
+
+If the final node receives an onion packet with the `incomplete_payment` flag
+set, then the payment MAY be a "base" atomic multipath payment.
+Such "base" atomic multipath payments will use the same `payment_hash` for all paths.
+
+The `amt_to_forward` value will be the actual total amount that the final node
+should expect.
+Thus, in a multipath payment, the incoming HTLC `incoming_htlc_amt` will be lower
```
[the remark this answers is not in the document]
Comment (ZmnSCPxj) [reply 04-onion-routing.md]: I have come to agree.

File: 04-onion-routing.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (Roasbeef) [reply 04-onion-routing.md]: Which users now this as AMP? When the majority of those that are aware of LN, here "AMP", they think of the original ML post. The only individuals that know of this partial payment feature are those that were at the summit.

[the remark this answers is not in the document]
Comment (ZmnSCPxj) [reply 04-onion-routing.md]: Okay, I shall change it then.

[the remark this answers is not in the document]
Comment (ZmnSCPxj) [reply 04-onion-routing.md]: "ZKCP-preseerving proper multipath payments"

**Pull request closed without merging**